### PR TITLE
Core/Bank: Reagent bank

### LINKED ISSUE #16077: [master] Core/Reagent Bank - Not Yet Implemented
Tried to purchase Reagent bank tab in Orgrimmar bank. Nothing happened and I had plenty of gold.

Will try in other cities with Alliance faction as well. Just bringing it to light.

TrinityCore rev. 4695e23 2015-12-09 01:33:37 +0100 (6.x branch) (Unix, Release)

### PR DESCRIPTION
**Pull request by aquadeus**
**Changes proposed:**

-  Implement reagent bank opcodes 

**Issues addressed:**

Closes #16077

**Tests performed:**

Tested in-game

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -10604,7 +10622,12 @@ InventoryResult Player::CanStoreItem_InSpecificSlot(uint8 bag, uint8 slot, ItemP
         if (bag == INVENTORY_SLOT_BAG_0)
         {
             // prevent cheating
-            if ((slot >= BUYBACK_SLOT_START && slot < BUYBACK_SLOT_END) || slot >= PLAYER_SLOT_END)
+            if ((slot >= BUYBACK_SLOT_START && slot < BUYBACK_SLOT_END) || slot >= PLAYER_SLOT_END
+                || (IsReagentBankPos(bag, slot) && !IsReagentBankUnlocked()))
+                return EQUIP_ERR_WRONG_BAG_TYPE;
+
+            // can't store anything else than crafting reagents in Reagent Bank
+            if (IsReagentBankPos(bag, slot) && !pProto->IsCraftingReagent())
```
[the remark this answers is not in the document]
Comment (aquadeus) [reply src/server/game/Entities/Player/Player.cpp:10630]: done ty 😄

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -1362,7 +1365,7 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         InventoryResult CanEquipUniqueItem(ItemTemplate const* itemProto, uint8 except_slot = NULL_SLOT, uint32 limit_count = 1) const;
         InventoryResult CanUnequipItems(uint32 item, uint32 count) const;
         InventoryResult CanUnequipItem(uint16 src, bool swap) const;
-        InventoryResult CanBankItem(uint8 bag, uint8 slot, ItemPosCountVec& dest, Item* pItem, bool swap, bool not_loading = true) const;
+        InventoryResult CanBankItem(uint8 bag, uint8 slot, ItemPosCountVec& dest, Item* pItem, bool swap, bool not_loading = true, bool reagent = false) const;
```
[the remark this answers is not in the document]
Comment (aquadeus) [reply src/server/game/Entities/Player/Player.h]: done

File: src/server/game/Handlers/BankHandler.cpp [hunk 3]
```diff
@@ -150,6 +150,152 @@ void WorldSession::HandleBuyBankSlotOpcode(WorldPackets::Bank::BuyBankSlot& pack
...
+    }
+
+    if (_player->IsReagentBankUnlocked())
+    {
+        TC_LOG_DEBUG("network", "WORLD: HandleBuyReagentBankOpcode - Player (%s, name: %s) tried to unlock reagent bank a 2nd time.", _player->GetGUID().ToString().c_str(), _player->GetName().c_str());
+        return;
+    }
+
+    int64 price = 100 * GOLD;
+    _player->ModifyMoney(-int64(price));
```
[the remark this answers is not in the document]
Comment (aquadeus) [reply src/server/game/Handlers/BankHandler.cpp]: done

File: src/server/game/Entities/Player/Player.cpp [hunk 4]
```diff
@@ -10080,6 +10080,20 @@ uint32 Player::GetItemCountWithLimitCategory(uint32 limitCategory, Item* skipIte
     return count;
 }
 
+std::vector<Item*> Player::GetCraftingReagentItems()
+{
+    std::vector<Item*> itemList = std::vector<Item*>();
+    ForEachItem(ItemSearchLocation::Equipment | ItemSearchLocation::Inventory, [&itemList](Item* item)
```
Comment (mdX7) [inline src/server/game/Entities/Player/Player.cpp:10086]: can we equip crafting reagents?

File: src/server/game/Handlers/BankHandler.cpp [hunk 5]
```diff
@@ -150,6 +150,152 @@ void WorldSession::HandleBuyBankSlotOpcode(WorldPackets::Bank::BuyBankSlot& pack
...
+    if (!_player->IsReagentBankUnlocked())
+    {
+        _player->SendEquipError(EQUIP_ERR_REAGENT_BANK_LOCKED);
+        return;
+    }
+
+    // query all reagents from player's inventory
+    std::vector<Item*> items = _player->GetCraftingReagentItems();
+
+    if (!items.empty())
```
Comment (mdX7) [inline src/server/game/Handlers/BankHandler.cpp:189]: you mustn't explicitly check if its empty if you are just iterating it. For loops are essentially while loops which check the condition before any action is done, and if the condition isnt satisfied the actions are not executed. That means if we have an empty vector the actions within the loop arent executed.

File: src/server/game/Entities/Player/Player.cpp (hunk 4 above)
Comment (aquadeus) [reply src/server/game/Entities/Player/Player.cpp:10086]: you are right idk why i was thinking about enchantment rods equipped

File: src/server/game/Handlers/BankHandler.cpp (hunk 5 above)
Comment (aquadeus) [reply src/server/game/Handlers/BankHandler.cpp:189]: ty 😄

**Comment by Shauren** [comment]
Changes
* Items in reagent bank can be used to complete quests
* Add missing gold check for unlocking

**Pull request merged**